transfer/custom.go

```diff
@@ -0,0 +1,376 @@
+	return &customAdapterTransferRequest{"download", oid, size, "", action}
+}
+
+type customAdapterTerminateRequest struct {
+	MessageType string `json:"type"`
+}
+
+func NewCustomAdapterTerminateRequest() *customAdapterTerminateRequest {
+	return &customAdapterTerminateRequest{"terminate"}
+}
+
+// A common struct that allows all types of response to be identified
+type customAdapterResponseMessage struct {
+	Id             string           `json:"id"`
+	Error          *api.ObjectError `json:"error"`
+	Oid            string           `json:"oid"`
+	Path           string           `json:"path,omitempty"` // always blank for upload
+	BytesSoFar     int64            `json:"bytesSoFar"`
+	BytesSinceLast int              `json:"bytesSinceLast"`
+}
+
+func (a *customAdapter) Begin(maxConcurrency int, cb TransferProgressCallback, completion chan TransferResult) error {
+	// If config says not to launch multiple processes, downgrade incoming value
+	useConcurrency := maxConcurrency
+	if !a.concurrent {
+		useConcurrency = 1
+	}
+	a.originalConcurrency = maxConcurrency
+
+	tracerx.Printf("xfer: Custom transfer adapter %q using concurrency %d", a.name, useConcurrency)
+
+	// Use common workers impl, but downgrade workers to number of processes
+	return a.adapterBase.Begin(useConcurrency, cb, completion)
+}
+
+func (a *customAdapter) ClearTempStorage() error {
+	// no action requred
+	return nil
+}
+
+func (a *customAdapter) WorkerStarting(workerNum int) (interface{}, error) {
+
+	// Start a process per worker
+	// If concurrent = false we have already dialled back workers to 1
+	tracerx.Printf("xfer: starting up custom transfer process %q for worker %d", a.name, workerNum)
+	cmd := subprocess.ExecCommand(a.path, a.args)
+	outp, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get stdout for custom transfer command %q remote: %v", a.path, err)
+	}
+	inp, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get stdin for custom transfer command %q remote: %v", a.path, err)
+	}
+	// Capture stderr to trace
+	tracer := &traceWriter{}
+	tracer.processName = filepath.Base(a.path)
+	cmd.Stderr = tracer
+	err = cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to start custom transfer command %q remote: %v", a.path, err)
+	}
+	// Set up buffered reader/writer since we operate on lines
+	ctx := &customAdapterWorkerContext{workerNum, cmd, outp, bufio.NewReader(outp), inp, tracer}
+
+	// send initiate message
+	initReq := NewCustomAdapterInitRequest(a.getOperationName(), a.concurrent, a.originalConcurrency)
+	resp, err := a.exchangeMessage(ctx, initReq)
+	if err != nil {
+		a.abortWorkerProcess(ctx)
+		return nil, err
+	}
+	if resp.Error != nil {
+		a.abortWorkerProcess(ctx)
+		return nil, fmt.Errorf("Error initializing custom adapter %q worker %d: %v", a.name, workerNum, resp.Error.Error())
```

Comment:
Don't think we need to call `Error()` here specifically, `%v` should figure it out. It's also consistent with the error formatting above.